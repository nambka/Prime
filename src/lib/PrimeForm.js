import React from 'react';

// Render the input form to enter a number
class PrimeForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      inputNum: '',
      answer: ''
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e) {
    this.setState({
      inputNum: e.target.value
    });
  }

  handleSubmit(e) {
    // alert('Submitted data: ' + this.state.inputNum);
    e.preventDefault();
    const num = this.state.inputNum
    
    // // Naive way to check a prime
    // function isPrime(number) {
    //   if(num < 2) return false;

    //   for (let k = 2; k < num; k++){
    //     if( num % k == 0){
    //       return false;
    //     }
    //   }
    //   return true;
    // }

    function isPrime(number) {
      if (number < 2) return false;
      if (number % 2 == 0) return (number == 2);
      const sqrt = Math.sqrt(number);
      // var i;
      for (let i = 3; i <= sqrt; i += 2) {
          if (number % i == 0) return false;
      }
      return true;
    }

    function findPrimeLoop(oddNumber) {
      while (!isPrime(oddNumber)) {
        oddNumber -= 2;
      }
      return oddNumber
    }

    // Find prime < input number
    function calcPrime(number) {
      var num = parseInt(number);

      if (num < 0) {
        alert('Please input a positive integer!');
        return ''
      }

      if ((num == 0) || (num == 1) || (num == 2)) {
        return 'None'
      }

      if (num == 3) {
        return '2'
      }

      if (num > Number.MAX_SAFE_INTEGER) {
        return 'Input a number less than 2^53 - 1'
      }
      
      if (num % 2 == 0) {
        num -= 1;
        if (isPrime(num)) {
          return num
        } else {
          return findPrimeLoop(num)
        }
      } else {
        if (isPrime(num)) num -= 2;
        return findPrimeLoop(num)
      }


      // Benchmark
      // Freezing browser
      // Failsafe -1 0 1 2 2.2 10e4(99991)

      // Sieve: https://www.geeksforgeeks.org/nearest-prime-less-given-number-n/
      // https://stackoverflow.com/questions/3885817/how-do-i-check-that-a-number-is-float-or-integer/20779354#20779354

      // Number.MAX_SAFE_INTEGER: (2^53 - 1) = 9,007,199,254,740,992 / 9007199254740992

      // Number.isInteger(123) //true
      // Number.isInteger(-123) //true
      // Number.isInteger(5-2) //true
      // Number.isInteger(0) //true
      // Number.isInteger(0.5) //false
      // Number.isInteger('123') //false
      // Number.isInteger(false) //false
      // Number.isInteger(Infinity) //false
      // Number.isInteger(-Infinity) //false
      // Number.isInteger(0 / 0) //false
    }

    // // Find prime <= input number
    // function calcPrime(number) {
    //   var num = parseInt(number)

    //   if (num < 0) {
    //     alert('Please input a positive integer!');
    //     return ''
    //   }

    //   if ((num == 0) || (num == 1) || (num == 2)) {
    //     return 'None'
    //   }

    //   if (num > Number.MAX_SAFE_INTEGER) {
    //     return 'Input a number less than 2^53 - 1'
    //   }
      
    //   if (num % 2 == 0) {
    //     num -= 1;
    //     if (isPrime(num)) {
    //       return num
    //     } else {
    //       return findPrimeLoop(num)
    //     }
    //   } else {
    //     return findPrimeLoop(num)
    //   }
    // }

    this.setState({
      answer: calcPrime(num)
    })    
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit} className="container-fluid nambk-container-input text-center">
        <h2 className="nambk-header">Prime Finder</h2>
        <div className="nambk-header-desc">Highest Prime Number lower than the Input Number</div>

        <label>Input a number:
          <input type="number" value={this.state.inputNum} onChange={this.handleChange} placeholder="Enter a number"/>
        </label>
        <input type="submit" value="Go"/><br/>
        <label>Highest prime #:
          <input type="text" value={this.state.answer} readOnly />   
        </label>
      </form>
    );
  }
}
export default PrimeForm;