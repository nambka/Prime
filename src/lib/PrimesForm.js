import React from 'react';

// Render the input form to enter two numbers
class PrimesForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      inputFirstNum: 1,
      inputLastNum: 1000000,
      answer: ['Primes']
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e) {
    this.setState({
      [e.target.name]: e.target.value,
    });
  }

  handleSubmit(e) {
    e.preventDefault();
    const num1 = Number(this.state.inputFirstNum)
    const num2 = Number(this.state.inputLastNum)
    const max = num1 > num2 ? num1 : num2
    const min = max === num1 ? num2 : num1

    function sieve(limit, initial) {
      var bools = [];
      var primes = [];
      for (var i = 1; i < limit; i++) { bools.push(true); }
      for (var i = 1; i <= Math.sqrt(limit); i++) {
        if (bools[i-2]) {
          for (var j = i*2; j <= limit; j += i) {
            bools[j-2] = false;
          }
        }
      }
      for (var i = 0; i < initial-2; i++) { bools[i] = false; }
      for (var p = 0; p < bools.length; p++) {
        if (bools[p]) { primes.push(p+2); }
      }
      return primes.length === 0 ? ['None in Range'] : primes
    }

    this.setState({
      answer: sieve(max, min),
    });   
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit} className="container-fluid nambk-container-input">
        <h2 className="nambk-header">Prime Numbers</h2>
        <div className="nambk-header-desc">Between Two Numbers</div>

        <input type="number" min="1" max="9999999" value={this.state.inputFirstNum} onChange={this.handleChange} id="inputFirstNum" name="inputFirstNum" placeholder="Enter a number"/>
        <input type="number" min="1" max="9999999" value={this.state.inputLastNum} onChange={this.handleChange} id="inputLastNum" name="inputLastNum" placeholder="Enter a number"/>
        <input type="submit" value="Go"/>
        <div className="mt-3" style={{
            backgroundColor: '#cde2', 
            padding: '5px 10px', 
            width: '100%', 
            flex: '1', 
            boxSizing: 'border-box',
            minHeight: '60px', 
            borderRadius: '7px', 
          }}>
          {this.state.answer.join(', ')}
        </div>   
      </form>
    );
  }
}
export default PrimesForm;