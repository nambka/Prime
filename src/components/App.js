import React from 'react';
import '../assets/css/nav.css';
import '../assets/css/style.css';
import Header from './Header';
import Footer from './Footer';
import PrimeForm from '../lib/PrimeForm';
import PrimesForm from '../lib/PrimesForm';

import { ThemeProvider } from 'styled-components';
import themes from '../service/theme';
import { useDarkMode } from '../lib/DarkMode';
import GlobalStyle from '../service/theme/global';

const App = () => {
  const [theme, toggleTheme] = useDarkMode();

  return (
    <ThemeProvider theme={theme === 'light' ? themes.lightTheme : (theme === 'dark' ? themes.darkTheme : themes.moonTheme)}>
      <GlobalStyle />
      <Header />
      <main className="container-fluid text-center">
        <PrimeForm />
        <PrimesForm />
      </main>
      <Footer theme={theme} toggleTheme={toggleTheme}/>
    </ThemeProvider> 
  );
};
export default App;