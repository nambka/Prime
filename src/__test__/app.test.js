import React from 'react';
import App from '../components/app';
import Header from '../components/header';
import Footer from '../components/header';
import PrimeForm from '../lib/PrimeForm';
import PrimesForm from '../lib/PrimesForm';
import { shallow, mount } from 'enzyme';
import TestRenderer from 'react-test-renderer';

describe('App:', () => {
  it('renders components without errors', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find(Header)).toHaveLength(1);
    expect(wrapper.find('.container-fluid')).toHaveLength(1);
    expect(wrapper.find(PrimeForm)).toHaveLength(1);
    expect(wrapper.find(PrimesForm)).toHaveLength(1);
    expect(wrapper.find(Footer)).toHaveLength(1); // for Jest
    // expect(wrapper.find(Foo)).to.have.length(3); // for Chai Assertion Library
    // expect(wrapper.find(Foo).length).toBe(1); // for Enzyme
    expect(wrapper).toMatchSnapshot();
    // console.log(wrapper.debug());
  });

  it('snapshot renders', () => {
    const component = TestRenderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

});