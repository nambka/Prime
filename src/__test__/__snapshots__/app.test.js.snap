// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App: renders components without errors 1`] = `ShallowWrapper {}`;

exports[`App: snapshot renders 1`] = `
<div>
  <nav
    aria-label="Local"
    className="js touch css-sticky ac-ln-sticking"
    dir="ltr"
    id="ac-localnav"
    lang="en-US"
    role="navigation"
  >
    <div
      className="ac-ln-wrapper"
    >
      <div
        className="ac-ln-background"
      />
      <div
        className="ac-ln-content"
      >
        <div
          className="ac-ln-title"
        >
          <a
            className="border-radius-50"
            href="/nambk"
          >
            <img
              alt="logo"
              src="logo.png"
            />
          </a>
        </div>
        <div
          className="ac-ln-menu"
        >
          <a
            className="ac-ln-menucta-anchor ac-ln-menucta-anchor-open"
            href="/nambk"
            id="ac-ln-menustate-open"
          >
            <span
              className="ac-ln-menucta-anchor-label"
            >
              Local Nav Open Menu
            </span>
          </a>
          <a
            className="ac-ln-menucta-anchor ac-ln-menucta-anchor-close"
            href="/nambk"
            id="ac-ln-menustate-close"
          >
            <span
              className="ac-ln-menucta-anchor-label"
            >
              Local Nav Close Menu
            </span>
          </a>
          <div
            className="ac-ln-menu-tray"
          >
            <ul
              className="ac-ln-menu-items"
            >
              <li
                className="ac-ln-menu-item"
              >
                <span
                  aria-current="page"
                  aria-disabled="true"
                  className="ac-ln-menu-link current"
                  role="link"
                >
                  Home
                </span>
              </li>
              <li
                className="ac-ln-menu-item"
              >
                <a
                  className="ac-ln-menu-link"
                  href="/nambk"
                >
                  Design
                </a>
              </li>
              <li
                className="ac-ln-menu-item"
              >
                <a
                  className="ac-ln-menu-link"
                  href="/nambk"
                >
                  Products
                </a>
              </li>
            </ul>
          </div>
          <div
            className="ac-ln-actions"
          >
            <div
              aria-hidden="true"
              className="ac-ln-action ac-ln-action-menucta"
            >
              <label
                className="ac-ln-menucta"
                htmlFor="ac-ln-menustate"
              >
                <span
                  className="ac-ln-menucta-chevron"
                />
              </label>
            </div>
            <div
              className="ac-ln-action ac-ln-action-button"
            >
              <a
                className="ac-ln-button mr-3"
                href="/nambk"
              >
                Login 
              </a>
              <button
                className="btn btn-sm btn-icon btn-soft-light"
                id="btnDarkMode"
                onClick={[Function]}
              >
                  
                <i
                  className="fa fa-moon-o"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container-fluid text-center"
  >
    <form
      className="container-fluid nambk-container-input text-center"
      id="frmFindHighestPrime"
      onSubmit={[Function]}
    >
      <h2
        className="nambk-header"
      >
        Highest Prime
      </h2>
      <div
        className="nambk-header-desc"
      >
        lower than Input Number
      </div>
      <div>
        <input
          className="nambk-input"
          id="inputNumber"
          name="inputNumber"
          onChange={[Function]}
          placeholder="Enter a number"
          style={
            Object {
              "width": "50%",
            }
          }
          type="number"
          value=""
        />
        <input
          className="nambk-btn nambk-btn-primary"
          id="btnFindHighestPrime"
          type="submit"
          value="Go"
        />
      </div>
      <div
        className="mt-3"
        id="result"
        readOnly={true}
        style={
          Object {
            "minHeight": "2rem",
          }
        }
      >
        
      </div>
    </form>
    <form
      className="container-fluid nambk-container-input"
      id="frmFindPrimes"
      onSubmit={[Function]}
    >
      <h2
        className="nambk-header"
      >
        Prime Numbers
      </h2>
      <div
        className="nambk-header-desc"
      >
        Between Two Numbers
      </div>
      <div>
        <input
          className="nambk-input"
          id="inputFirstNum"
          max="9999999"
          min="1"
          name="inputFirstNum"
          onChange={[Function]}
          placeholder="Enter a number"
          style={
            Object {
              "width": "33%",
            }
          }
          type="number"
          value={1}
        />
        <input
          className="nambk-input"
          id="inputLastNum"
          max="9999999"
          min="1"
          name="inputLastNum"
          onChange={[Function]}
          placeholder="Enter a number"
          style={
            Object {
              "width": "33%",
            }
          }
          type="number"
          value={1000}
        />
        <input
          className="nambk-btn nambk-btn-primary"
          id="btnFindPrimes"
          type="submit"
          value="Go"
        />
      </div>
      <div
        className="mt-3"
        id="result"
        style={
          Object {
            "minHeight": "4rem",
          }
        }
      >
        Primes
      </div>
    </form>
  </div>
  <footer
    className="gradient-half-primary-body-v1 py-4"
    style={
      Object {
        "height": "calc(100% - 60px)",
      }
    }
  >
    <div
      className="container"
    >
      <div
        className="row justify-content-md-between font-size-1 py-3"
        style={
          Object {
            "backgroundImage": "url(world-map-white.png)",
            "backgroundPositionX": "0px",
            "backgroundPositionY": "30%",
            "backgroundRepeat": "no-repeat",
          }
        }
      >
        <div
          className="col-lg-3 mb-4 mb-lg-0"
        >
          <div
            className="d-flex h-100 flex-column"
          >
            <div
              className="d-flex align-items-center mb-4"
            >
              <img
                alt="Landing Page Logo"
                src="snapseed256.png"
                width="32"
              />
              <span
                className="h5 mb-0 ml-3"
              >
                Powered by Nambk
              </span>
            </div>
            <p
              className="font-size-0-75"
            >
              A brand new journey to the beautiful angles of this world. Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit...
            </p>
            <div
              className="d-flex align-items-center"
            >
              <a
                className="btn btn-xss btn-soft-light mr-2"
                href="/settings"
              >
                <i
                  className="fa fa-cogs mr-1"
                />
                Preferences
              </a>
              <button
                className="btn btn-sm btn-icon btn-soft-light"
                data-original-title=""
                data-title="Day/Night Mode"
                data-toggle="tooltip"
                id="btnDarkMode"
                title=""
              >
                <i
                  className="fa fa-moon-o"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </footer>
</div>
`;
